### DISCUSSION TITLE
OAuth2 Provider

### DISCUSSION THREAD

**Discussion by SGarno** [Dan6erbond/sk-auth#31] [Q&A] [answered]
If I understand your documentation correctly, the provider parameter in the profile is supposed to determine the path to route to in order to run that provider configuration.

For example:

```javascript
        new OAuth2Provider({
            scope: "https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email",
            params: { grant_type: "authorization_code" },
            accessTokenUrl: "https://accounts.google.com/o/oauth2/token",
            requestTokenUrl: "https://accounts.google.com/o/oauth2/auth",
            authorizationUrl: "https://accounts.google.com/o/oauth2/auth?response_type=code",
            profileUrl: "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
            profile (profile) {
                return { ...profile, provider: "custom" };
            },
            clientId: "",
            clientSecret: ""
        })
```

Should allow me to route to /api/auth/signin/custom and it will perform the custom authentication (google in this case).   Correct?

I am getting this error:

` GET http://localhost:3000/api/auth/signin/custom 404 (Not Found)`

Did I miss something in the docs?

**Comment by Dan6erbond** [comment, marked as answer]
Sorry! That's a bit confusing currently in the demo app I use the `profile()` callback simply to remap the profile and add the `provider` information for later use in the `jwt()` callback. As you can see, I'm able to append as many connections as I want then allowing me to display the connections in the UI:

```ts
export const appAuth = new SvelteKitAuth({
  callbacks: {
    jwt(token, profile) {
      if (profile?.provider) {
        const { provider, ...account } = profile; // profile returned by profile() callback
        token = {
          ...token,
          user: {
            ...(token.user ?? {}),
            connections: { ...(token.user?.connections ?? {}), [provider]: account },
          },
        };
      }

      return token;
    },
});
```

SvelteKitAuth internally uses the `id` field in the provider configs to map the routes. Due to issues with typing it's an optional param, but must be set if you're creating your own provider like so:

```ts
new OAuth2Provider({
  id: "custom",
  scope:
    "https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email",
  params: { grant_type: "authorization_code" },
  accessTokenUrl: "https://accounts.google.com/o/oauth2/token",
  requestTokenUrl: "https://accounts.google.com/o/oauth2/auth",
  authorizationUrl:
    "https://accounts.google.com/o/oauth2/auth?response_type=code",
  profileUrl: "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
  profile(profile) {
    return { ...profile, provider: "custom" };
  },
  clientId: "",
  clientSecret: "",
});
```

This will map the routes to `/api/auth/signin/custom` and `/api/auth/callback/custom`. 🙂

**Comment by SGarno** [comment]
Brilliant!   I knew it was something simple I was missing.    Thanks.
